drivers/storage/ide/pciidex/pciidex.c

```diff
@@ -11,10 +11,327 @@
+                            FALSE,
+                            &Fdo);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT("Failed to create the FDO with status 0x%08lx\n", Status);
+        return Status;
+    }
+
+    FdoExtension = Fdo->DeviceExtension;
+
+    RtlZeroMemory(FdoExtension, sizeof(FDO_DEVICE_EXTENSION));
+    FdoExtension->Common.IsFDO = TRUE;
+    FdoExtension->Common.Self = Fdo;
+    FdoExtension->DriverObject = DriverObject;
+    FdoExtension->ControllerNumber = ControllerNumber++;
+
+    KeInitializeSpinLock(&FdoExtension->BusDataLock);
+    KeInitializeEvent(&FdoExtension->DeviceSyncEvent, SynchronizationEvent, TRUE);
+
+    Status = IoAttachDeviceToDeviceStackSafe(Fdo, PhysicalDeviceObject, &FdoExtension->Ldo);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT("Failed to attach the FDO with status 0x%08lx\n", Status);
+        return Status;
+    }
+
+    /* DMA buffers alignment */
+    Fdo->AlignmentRequirement = FdoExtension->Ldo->AlignmentRequirement;
+    if (Fdo->AlignmentRequirement < FILE_WORD_ALIGNMENT)
+        Fdo->AlignmentRequirement = FILE_WORD_ALIGNMENT;
+
+    Status = PciIdeXQueryInterface(FdoExtension,
+                                   &GUID_BUS_INTERFACE_STANDARD,
+                                   &FdoExtension->BusInterface,
+                                   sizeof(BUS_INTERFACE_STANDARD));
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("No bus interface\n");
+        goto Failure;
+    }
+
+    Status = PciIdeXGetConfigurationInfo(FdoExtension);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Unable to retrieve the configuration info\n");
+        goto Failure;
+    }
+
+    Fdo->Flags &= ~DO_DEVICE_INITIALIZING;
+
+    return STATUS_SUCCESS;
+
+Failure:
+    if (FdoExtension)
+    {
+        if (FdoExtension->Ldo)
+            IoDetachDevice(FdoExtension->Ldo);
+    }
+
+    if (Fdo)
+    {
+        IoDeleteDevice(Fdo);
+    }
+
+    return Status;
+}
+
+static
+CODE_SEG("PAGE")
+VOID
+PciIdeXCreateIdeDirectory(VOID)
+{
+    HANDLE Handle;
+    NTSTATUS Status;
+    OBJECT_ATTRIBUTES ObjectAttributes;
+    UNICODE_STRING DirectoryName = RTL_CONSTANT_STRING(L"\\Device\\Ide");
+
+    PAGED_CODE();
+
+    InitializeObjectAttributes(&ObjectAttributes,
+                               &DirectoryName,
+                               OBJ_CASE_INSENSITIVE | OBJ_PERMANENT | OBJ_KERNEL_HANDLE,
+                               NULL,
+                               NULL);
+
+    Status = ZwCreateDirectoryObject(&Handle, DIRECTORY_ALL_ACCESS, &ObjectAttributes);
+    if (NT_SUCCESS(Status))
+    {
+        PVOID DirectoryObject;
+
+        /* We don't need the handle */
+        ObReferenceObjectByHandle(Handle,
+                                  FILE_READ_ATTRIBUTES,
+                                  NULL,
+                                  KernelMode,
+                                  &DirectoryObject,
+                                  NULL);
+        ZwClose(Handle);
+    }
```

Comment:
Shouldn't you be handling this gracefully if `ZwCreateDirectoryObject` fails, if ever? Or are you surely confident in yourself that this will never fail?

Reply:
Added a comment